Ensure n-ary to_index inlines

mbauman opened the pull request:
While I had fixed the type instability of to_index in https://github.com/JuliaLang/julia/commit/6a3c173633a8c019177daad7d752f0aa69bd621d, it stopped inlining even sooner!

This is the right solution — use a recursive n-ary splatted argument list.  This fixes #11819 properly.  I don't have time to run the test suite locally, but assuming CI passes this is good to go.

mbauman committed:
Ensure to_index inlines by separating the n-ary version

This separates the n-ary version of to_index to to_indexes. Recursive definitions with tuples don't inline like recursive splatted arguments.

mbauman commented:
Argh, it seems Win64 still doesn't like this simplification to the `Array` getindex methods.

tkelman commented:
The win64 failure was the timeout that's been happening everywhere (and the latest commit here more-or-less worked for me locally), but I didn't try locally on win32 yet and there's apparently an assertion failure in llvm
```
This application has requested the Runtime to terminate it in an unusual way.
Please contact the application's support team for more information.
Assertion failed!
Program: C:\projects\julia\usr\bin\julia.exe
File: /home/Tony/julia32/deps/llvm-3.3/include/llvm/ADT/SmallVector.h, Line 144
Expression: begin() + idx < end()
	From worker 2:	     * math                Worker 2 terminated.
```

mbauman committed:
Deprecate to_index(::Tuple)

While this was an internal function, it was used in at least a few other packages.  It is easy to deprecate, so we may as well do so.

mbauman commented:
Well, lets see if it happens again.  I squashed and added a deprecation (since a few packages had been using this and it was easy to deprecate).

tkelman commented:
(initially posted this in the wrong issue, don't triage before coffee)

looks like the win32 problem is legitimate, though different error this time?
```
exception on 2: ERROR: LoadError: test failed: !(isequal(s.dict.keys,k))
 in expression: !(isequal(s.dict.keys,k))
 in error at error.jl:21
 in default_handler at test.jl:29
 in do_test at test.jl:52
 in include at boot.jl:254
 in runtests at C:\projects\julia\test\testdefs.jl:197
 in anonymous at multi.jl:837
 in run_work_thunk at multi.jl:590
 in anonymous at task.jl:837
while loading C:\projects\julia\test\sets.jl, in expression starting on line 95
	From worker 3:	     * replutil             in  10.43 seconds
ERROR: LoadError: LoadError: test failed: !(isequal(s.dict.keys,k))
 in expression: !(isequal(s.dict.keys,k))
 in anonymous at task.jl:1403
while loading C:\projects\julia\test\sets.jl, in expression starting on line 95
while loading C:\projects\julia\test\runtests.jl, in expression starting on line 5
 
	From worker 2:	     * sets                Command exited with code 1
```

mbauman committed:
Simplify getindex and setindex! for Array
